server/api/src/rsDataObjLseek.cpp

```diff
@@ -10,112 +7,102 @@
-                 "rsDataObjLseek: l1descInx %d out of range",
-                 l1descInx );
+    if (l1descInx <= 2 || l1descInx >= NUM_L1_DESC) {
+        rodsLog(LOG_ERROR, "%s: l1descInx %d out of range", __func__, l1descInx);
         return SYS_FILE_DESC_OUT_OF_RANGE;
     }
-    if ( L1desc[l1descInx].inuseFlag != FD_INUSE ) {
+
+    auto& l1desc = L1desc[l1descInx];
+
+    if (l1desc.inuseFlag != FD_INUSE) {
         return BAD_INPUT_DESC_INDEX;
     }
-    if ( L1desc[l1descInx].remoteZoneHost != NULL ) {
-        /* cross zone operation */
-        dataObjLseekInp->l1descInx = L1desc[l1descInx].remoteL1descInx;
-        status = rcDataObjLseek( L1desc[l1descInx].remoteZoneHost->conn,
-                                 dataObjLseekInp, dataObjLseekOut );
+
+    if (l1desc.remoteZoneHost) {
+        // Cross zone operation.
+        dataObjLseekInp->l1descInx = l1desc.remoteL1descInx;
+        const auto ec = rcDataObjLseek(l1desc.remoteZoneHost->conn, dataObjLseekInp, dataObjLseekOut);
         dataObjLseekInp->l1descInx = l1descInx;
-        return status;
+        return ec;
     }
 
-    l3descInx = L1desc[l1descInx].l3descInx;
+    const int l3descInx = l1desc.l3descInx;
 
-    if ( l3descInx <= 2 ) {
-        rodsLog( LOG_NOTICE,
-                 "rsDataObjLseek: l3descInx %d out of range",
-                 l3descInx );
+    if (l3descInx <= 2) {
+        rodsLog(LOG_ERROR, "%s: l3descInx %d out of range", __func__, l3descInx);
         return SYS_FILE_DESC_OUT_OF_RANGE;
     }
 
-    dataObjInfo = L1desc[l1descInx].dataObjInfo;
+    auto* dataObjInfo = l1desc.dataObjInfo;
 
-    // =-=-=-=-=-=-=-
-    // extract the host location from the resource hierarchy
+    // Extract the host location from the resource hierarchy.
     std::string location;
-    irods::error ret = irods::get_loc_for_hier_string( dataObjInfo->rescHier, location );
-    if ( !ret.ok() ) {
-        irods::log( PASSMSG( "rsDataObjLseek - failed in get_loc_for_hier_string", ret ) );
-        return -1;
+    if (const auto ret = irods::get_loc_for_hier_string(dataObjInfo->rescHier, location); !ret.ok()) {
+        irods::log(PASSMSG("rsDataObjLseek: failed in get_loc_for_hier_string", ret));
+        return ret.code();
     }
 
-
-    if ( getStructFileType( dataObjInfo->specColl ) >= 0 ) {
-        subStructFileLseekInp_t subStructFileLseekInp;
-        memset( &subStructFileLseekInp, 0, sizeof( subStructFileLseekInp ) );
+    // If the replica was opened in read-only mode, then don't allow the client
+    // to seek past the data size in the catalog. This is necessary because the
+    // size of the replica in storage could be larger than the data size in the
+    // catalog.
+    //
+    // For all other modes, let the seek operation do what it normally does.
+    const auto offset = (O_RDONLY == (l1desc.dataObjInp->openFlags & O_ACCMODE))
+        ? std::min(dataObjLseekInp->offset, dataObjInfo->dataSize)
+        : dataObjLseekInp->offset;
```

Comment:
yes, this looks good

Reply:
I may need to define the integer type for std::min here.